fix(api): accept avatar URLs and use email for user creation

pedroccastro opened the pull request:
## What does this PR do?

This PR fixes two bugs in **API v2** user creation endpoint `POST /v2/organizations/{orgId}/users`:

**Bug 1 - avatarValidator rejecting URLs:**
- The validator only accepted base64 encoded images
- API documentation and all e2e tests show URL examples
- Users received empty `avatarValidator: ""` constraint error

**Bug 2 - User creation failing when username provided:**
- `createNewUsersConnectToOrgIfExists` requires a valid email
- Service was passing username instead of email when username was provided
- Error: `"Invite failed because {username} is not a valid email address"`

**Key Changes:**
- `avatarValidator.ts`: Accept HTTPS URLs (HTTP rejected for security) and base64 images
- `organizations-users-service.ts`: Always use email for user creation; username is applied via `updateOrganizationUser` after creation

## Visual Demo

#### **Before - Avatar URL rejected**

<img width="500" alt="avatar_error" src="https://github.com/user-attachments/assets/dc1512d0-2c0e-4927-9454-1f67f30b21d7" />

#### **Before - Username causing TRPCError**

<img width="500" alt="trpc_error" src="https://github.com/user-attachments/assets/0f67cc77-5252-4b9a-8aaa-f1bc7183c83d" />

#### **After - Full payload working**

<img width="500" alt="working" src="https://github.com/user-attachments/assets/f5a8661e-c9f9-4a0c-a235-3ca83bf256b3" />

## How should this be tested?

**Test Scenario 1 - Avatar URL accepted:**
1. `POST /v2/organizations/{orgId}/users`
2. Body: `{ "email": "test@example.com", "avatarUrl": "https://avatars.githubusercontent.com/u/583231?v=4" }`
3. **Expected:** User created successfully with avatarUrl

**Test Scenario 2 - Username with avatar:**
1. `POST /v2/organizations/{orgId}/users`
2. Body:
```json
{
  "email": "test@example.com",
  "username": "testuser",
  "avatarUrl": "https://avatars.githubusercontent.com/u/583231?v=4",
  "locale": "pt",
  "timeZone": "America/Sao_Paulo",
  "timeFormat": 24
}
```
3. **Expected:** User created with correct username and avatarUrl

**Test Scenario 3 - Minimal payload (regression):**
1. `POST /v2/organizations/{orgId}/users`
2. Body: `{ "email": "minimal@example.com" }`
3. **Expected:** User created successfully (existing behavior preserved)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code
- [ ] **N/A** I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). 
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

pedroccastro committed:
fix(api): fix user creation with avatarUrl and username

- avatarValidator: accept URLs in addition to base64 images
- organizations-users-service: use email for user creation instead of username

The avatarValidator was incorrectly rejecting valid URLs even though the API documentation shows URL examples and all e2e tests use URLs.

The user creation was failing when username was provided because createNewUsersConnectToOrgIfExists requires a valid email, not username. The username is correctly applied via updateOrganizationUser after creation.

Fixes user creation endpoint POST /v2/organizations/{orgId}/users

keithwillcode reviewed (changes requested):
Since logic is changing, we should have automated tests to verify the changes

supalarry reviewed (commented):
Review:

supalarry commented on apps/api/v2/src/modules/organizations/users/index/services/organizations-users-service.ts:
```diff
@@ -68,13 +68,11 @@ export class OrganizationsUsersService {
       await this.checkForUsernameConflicts(org.id, userCreateBody.username);
     }
 
-    const usernameOrEmail = userCreateBody.username ? userCreateBody.username : userCreateBody.email;
```
But then we can rename `usernameOrEmail` to `email`

pedroccastro replied:
> But then we can rename `usernameOrEmail` to `email`

Good observation @supalarry! The rename would make the code cleaner. I took a quick look and those functions are exported through platform-libraries and used across multiple areas. There are some coordination concerns since changing their signatures would touch API v2, tRPC, OAuth clients and DSync. I'd need to investigate

pedroccastro committed:
test(api): add tests for avatar validator fixes

  Add comprehensive test coverage for avatarValidator changes:

  - Unit tests: 18 scenarios covering valid/invalid URLs and base64 images
  - E2E tests: user creation with username + avatarUrl (URL and base64)
  - Security validation: reject unsafe protocols and malformed data
  - Error message verification for validation failures

  Addresses testing requirements from code review

hariombalhara reviewed (dismissed):
LGTM !!

pedroccastro committed:
fix(api): enhance avatar validator security per code review

  - Reject HTTP URLs, accept only HTTPS for security and browser compatibility
  - Reject empty strings and whitespace (use null to reset avatar)
  - Update validation message to clarify HTTPS requirement
  - Add test coverage for new security restrictions

 Addresses security feedback from code review regarding mixed content vulnerabilities and clearer API semantics for avatar reset

pedroccastro commented on apps/api/v2/src/modules/users/validators/avatarValidator.ts:
```diff
@@ -1,11 +1,31 @@
-    const regex = /^data:image\/[^;]+;base64,(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
-    return regex.test(avatarString);
+  validate(avatarString: string): boolean {
+    if (!avatarString) return true;
```
@hariombalhara Good point! I've updated the validator to reject empty strings. Now only explicit `null` or `undefined` will reset the avatar. Empty strings and whitespace-only strings now return validation errors

pedroccastro commented on apps/api/v2/src/modules/users/validators/avatarValidator.ts:
```diff
@@ -1,11 +1,31 @@
+  private isValidUrl(value: string): boolean {
+    try {
+      const url = new URL(value);
+      return url.protocol === "http:" || url.protocol === "https:";
```
@hariombalhara Also makes sense! I've updated the validator to only accept HTTPS URLs. This prevents mixed content issues and aligns with our security standards. The error message now explicitly mentions "HTTPS only" to guide users

hariombalhara commented on apps/api/v2/src/modules/users/validators/avatarValidator.ts:
```diff
@@ -1,11 +1,32 @@
-    const regex = /^data:image\/[^;]+;base64,(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
-    return regex.test(avatarString);
+  validate(avatarString: string): boolean {
+    if (avatarString === null || avatarString === undefined) return true;
```
nit: TS confirms at line 9 that avatarString can be string only.  
So, we just need  
if (!avatarString?.trim()) return false;

pedroccastro committed:
refactor(api): simplify avatarValidator null check

Remove redundant null/undefined check since @IsOptional decorator already skips validation for undefined values. Updated test mock to include @IsOptional to match real DTO usage

pedroccastro committed:
Merge branch 'main' into fix/avatar-validator-accept-url

pedroccastro committed:
Merge branch 'main' into fix/avatar-validator-accept-url

pedroccastro committed:
Merge branch 'main' into fix/avatar-validator-accept-url

keithwillcode committed:
Merge branch 'main' into fix/avatar-validator-accept-url
